features/mbedtls/targets/TARGET_STM/md5_alt.c

```diff
@@ -0,0 +1,158 @@
+        return;
+    mbedtls_zeroize( ctx, sizeof( mbedtls_md5_context ) );
+}
+
+void mbedtls_md5_clone( mbedtls_md5_context *dst,
+                        const mbedtls_md5_context *src )
+{
+    *dst = *src;
+}
+
+void mbedtls_md5_starts( mbedtls_md5_context *ctx )
+{
+    /* HASH IP initialization */
+    if (HAL_HASH_DeInit(&ctx->hhash_md5) != 0) {
+        // error found to be returned
+        return;
+    }
+
+    /* HASH Configuration */
+    ctx->hhash_md5.Init.DataType = HASH_DATATYPE_8B;
+    if (HAL_HASH_Init(&ctx->hhash_md5) != 0) {
+        // return error code
+        return;
+    }
+    st_md5_save_hw_context(ctx);
+}
+
+void mbedtls_md5_process( mbedtls_md5_context *ctx, const unsigned char data[ST_MD5_BLOCK_SIZE] )
+{
+    st_md5_restore_hw_context(ctx);
+    if (HAL_HASH_MD5_Accumulate(&ctx->hhash_md5, (uint8_t *)data, ST_MD5_BLOCK_SIZE) != 0) {
+        return; // Return error code here
+    }
+    st_md5_save_hw_context(ctx);
+}
+
+void mbedtls_md5_update( mbedtls_md5_context *ctx, const unsigned char *input, size_t ilen )
+{
+    size_t currentlen = ilen;
+    st_md5_restore_hw_context(ctx);
+
+    // store mechanism to accumulate ST_MD5_BLOCK_SIZE bytes (512 bits) in the HW
+    if (currentlen == 0){ // only change HW status is size if 0
+        if(ctx->hhash_md5.Phase == HAL_HASH_PHASE_READY) {
+          /* Select the MD5 mode and reset the HASH processor core, so that the HASH will be ready to compute
+             the message digest of a new message */
+          HASH->CR |= HASH_ALGOSELECTION_MD5 | HASH_CR_INIT;
+        }
+        ctx->hhash_md5.Phase = HAL_HASH_PHASE_PROCESS;
+    } else if (currentlen < (ST_MD5_BLOCK_SIZE - ctx->sbuf_len)) {
+        // only buffurize
+        memcpy(ctx->sbuf+ctx->sbuf_len, input, currentlen);
+        ctx->sbuf_len += currentlen;
+    } else {
+        // fill buffer and process it
+        memcpy(ctx->sbuf + ctx->sbuf_len, input, (ST_MD5_BLOCK_SIZE - ctx->sbuf_len));
+        currentlen -= (ST_MD5_BLOCK_SIZE - ctx->sbuf_len);
+        mbedtls_md5_process(ctx, ctx->sbuf);
+        // Process every input as long as it is %64 bytes, ie 512 bits
+        size_t iter = currentlen / ST_MD5_BLOCK_SIZE;
+        if (iter !=0) {
+            if (HAL_HASH_MD5_Accumulate(&ctx->hhash_md5, (uint8_t *)(input + ST_MD5_BLOCK_SIZE - ctx->sbuf_len), (iter * ST_MD5_BLOCK_SIZE)) != 0) {
+                return; // Return error code here
+            }
+        }
+        // sbuf is completely accumulated, now copy up to 63 remaining bytes
+        ctx->sbuf_len = currentlen % ST_MD5_BLOCK_SIZE;
+        if (ctx->sbuf_len !=0) {
+            memcpy(ctx->sbuf, input + ilen - ctx->sbuf_len, ctx->sbuf_len);
+        }
+    }
+    st_md5_save_hw_context(ctx);
+}
+
+void mbedtls_md5_finish( mbedtls_md5_context *ctx, unsigned char output[16] )
+{
+    st_md5_restore_hw_context(ctx);
+    if (ctx->sbuf_len > 0) {
+        if (HAL_HASH_MD5_Accumulate(&ctx->hhash_md5, ctx->sbuf, ctx->sbuf_len) != 0) {
```

Comment:
hello @andresag01 
You are right : MD5 / SHA1 and SHA256 are using the same HW block (_HASH_).
FYI with the new version, I only manage 512 bits blocks (to allow the context swap).  ctx->sbuf_len can now have a value up to **63**.

This is ok to have such sizes in the `mbedtls_md5_finish` function, because when calling the `HAL_DIGEST` after `Accumulate`, the hardware will handle automatic padding. (this is not inserting zeroes in the middle of a message, only at its end).

kind regards